README.md
@idio/core
===

[![npm version](https://badge.fury.io/js/%40idio%2Fcore.svg)](https://npmjs.org/package/@idio/core)

<a href="https://github.com/idio/core"><img src="https://raw.github.com/idiocc/core/master/images/logo.svg?sanitize=true" width="150" align="left"></a>

`@idio/core` is a _Koa2_-based web server with some pre-installed middleware which facilitates quick creation of a web server with the essential functionality, such as serving static files, compression, body parsing, _etc_. It also provides full JSDoc documentation of all options for completion in IDEs. Other components such as `@idio/database`, `@idio/route` and `@idio/jsx` allow to build more complex websites (to come).

```sh
yarn add -E @idio/core
```

<p align="center">
<a href="https://idio.cc"><img alt="Developer-Friendly Suggestions For Middleware" src="app2.gif"></a>
</p>

<p align="center"><a href="#table-of-contents"><img src=".documentary/section-breaks/0.svg?sanitize=true"></a></p>

## Table Of Contents

- [Table Of Contents](#table-of-contents)
- [API](#api)
- [`async core(middlewareConfig?: MiddlewareConfig, config?: Config): IdioCore`](#async-coremiddlewareconfig-middlewareconfigconfig-config-idiocore)
  * [`MiddlewareConfig`](#type-middlewareconfig)
  * [`Config`](#type-config)
  * [`IdioCore`](#type-idiocore)
- [Middleware Configuration](#middleware-configuration)
  * [Session](#session)
  * [File Uploads](#file-uploads)
  * [Cross-Site Request Forgery](#cross-site-request-forgery)
  * [Parse Body](#parse-body)
  * [Checking Auth](#checking-auth)
  * [Logging](#logging)
  * [Compression](#compression)
  * [Static Files](#static-files)
  * [CORS](#cors)
  * [Frontend](#frontend)
- [Custom Middleware](#custom-middleware)
- [Router Set-up](#router-set-up)
- [Copyright](#copyright)

<p align="center"><a href="#table-of-contents"><img src=".documentary/section-breaks/1.svg?sanitize=true"></a></p>

## API

The package is available by importing its default function:

```js
import idioCore from '@idio/core'
```

<p align="center"><a href="#table-of-contents"><img src=".documentary/section-breaks/2.svg?sanitize=true"></a></p>

## `async core(`<br/>&nbsp;&nbsp;`middlewareConfig?: MiddlewareConfig,`<br/>&nbsp;&nbsp;`config?: Config,`<br/>`): IdioCore`

The `@idio/core` accepts 2 arguments which are the middleware configuration object and server configuration object. It is possible to start the server without any configuration, however it will do nothing, therefore it is important to add some middleware configuration.

__<a name="type-middlewareconfig">`MiddlewareConfig`</a>__: Middleware configuration for the `idio` `core` server.

|    Name    |                                                                              Type                                                                               |                                                      Description                                                       |
| ---------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------- | ---------------------------------------------------------------------------------------------------------------------- |
| session    | <em>[SessionOptions](#type-sessionoptions)</em>                                                                                                                 | `session` options.                                                                                                     |
| multer     | <em>[MulterOptions](#type-multeroptions)</em>                                                                                                                   | `multer` options.                                                                                                      |
| csrf       | <em>[CSRFOptions](#type-csrfoptions)</em>                                                                                                                       | `csrf` options.                                                                                                        |
| bodyparser | <em>[BodyparserOptions](#type-bodyparseroptions)</em>                                                                                                           | `bodyparser` options.                                                                                                  |
| compress   | <em>[CompressOptions](#type-compressoptions)</em>                                                                                                               | `compress` options.                                                                                                    |
| checkauth  | <em>[CheckauthOptions](#type-checkauthoptions)</em>                                                                                                             | `checkauth` options.                                                                                                   |
| logger     | <em>[LoggerOptions](#type-loggeroptions)</em>                                                                                                                   | `logger` options.                                                                                                      |
| static     | <em>[StaticOptions](#type-staticoptions)</em>                                                                                                                   | `static` options.                                                                                                      |
| cors       | <em>[CorsOptions](#type-corsoptions)</em>                                                                                                                       | `cors` options.                                                                                                        |
| frontend   | <em><a href="#type-frontendoptions" title="Allows to serve front-end JS files and CSS as modules, including from node_modules folder.">FrontendOptions</a></em> | `frontend` options. If the option is specified, the middleware always will be used, i.e., no need to pass `use: true`. |

__<a name="type-config">`Config`</a>__: Server configuration object.

| Name |      Type       |              Description               |  Default  |
| ---- | --------------- | -------------------------------------- | --------- |
| port | <em>number</em> | The port on which to start the server. | `5000`    |
| host | <em>string</em> | The host on which to listen.           | `0.0.0.0` |

---

The return type contains the _URL_, _Application_ and _Router_ instances, and the map of configured middleware, which could then be [passed to the router](#router-set-up).

[`import('@goa/koa').Application`](https://github.com/idiocc/goa/blob/master/doc/TYPES.md#type-_goaapplication) __<a name="type-_goaapplication">`_goa.Application`</a>__: An instance of the Koa application.

[`import('@goa/koa').Middleware`](https://github.com/idiocc/goa/blob/master/doc/TYPES.md#type-_goamiddleware) __<a name="type-_goamiddleware">`_goa.Middleware`</a>__: An async middleware function.

[`import('koa-router').Router`](https://github.com/alexmingoia/koa-router#exp_module_koa-router--Router) __<a name="type-koa-routerrouter">`koa-router.Router`</a>__: An instance of the Koa router.

[`import('http').Server`](https://nodejs.org/api/http.html#http_class_http_server) __<a name="type-httpserver">`http.Server`</a>__: An instance of the Node's Server class.

__<a name="type-idiocore">`IdioCore`</a>__: An object containing the url and references to the app, router and middleware.

|    Name    |                                                          Type                                                           |                                             Description                                              |         Default         |
| ---------- | ----------------------------------------------------------------------------------------------------------------------- | ---------------------------------------------------------------------------------------------------- | ----------------------- |
| url        | <em>string</em>                                                                                                         | The url on which the server is accessible.                                                           | `http://localhost:5000` |
| app        | <em><a href="#type-_goaapplication" title="An instance of the Koa application.">_goa.Application</a></em>               | The `Koa` application.                                                                               | -                       |
| router     | <em>Router</em>                                                                                                         | The `koa-router` instance.                                                                           | -                       |
| server     | <em><a href="#type-httpserver" title="An instance of the Node's Server class.">http.Server</a></em>                     | The `http` server instance.                                                                          | -                       |
| middleware | <em>Object&lt;string, <a href="#type-_goamiddleware" title="An async middleware function.">_goa.Middleware</a>&gt;</em> | The map of configured middleware functions which could then be set up to be used on a certain route. | -                       |

---

To start the server, the async method needs to be called and passed the middleware and server configuration objects. For example, the following code will start a server which serves static files with enabled compression.

```js
import idioCore from '@idio/core'

const Server = async () => {
  const { url } = await idioCore({
    logger: {
      use: true,
    },
    static: {
      use: true,
      root: 'example/static',
      mount: '/static',
    },
    compress: {
      use: true,
      config: {
        threshold: 1024,
      },
    },
  }, {
    port: 8080,
  })
  console.log('File available at: %s/static/test.txt', url)
}
```
```
File available at: http://localhost:8080/static/test.txt
```

<p align="center"><a href="#table-of-contents"><img src=".documentary/section-breaks/3.svg?sanitize=true"></a></p>

## Middleware Configuration

The middleware can be configured according to the `MiddlewareConfig`. `@idio/core` comes with some installed middleware as dependencies to speed up the process of creating a web server. Moreover, any custom middleware which is not part of the bundle can also be specified here (see [Custom Middleware](#custom-middleware)).

Each middleware accepts the following properties:

|  Property  |                                                                                                                             Description                                                                                                                             | Default |
| ---------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------- |
| `use` | Whether to use this middleware for every request. If not set to `true`, the configured middleware function will be included in the `middleware` property of the returned object, which can then be passed to a router configuration (not part of the `@idio/core`). | `false` |
| `config` | Configuration object expected by the middleware constructor.                                                                                                                                                                                                        | `{}` |
| `...props` | Any additional specific properties (see individual middleware configuration).                                                                                                                                                                                       |         |



<p align="center"><a href="#table-of-contents"><img src=".documentary/section-breaks/4.svg?sanitize=true" width="15"></a></p>
<a href="https://github.com/koajs/session"><img src="https://raw.github.com/idiocc/core/master/images/session.svg?sanitize=true" align="left" height="100"></a>
<details>
<summary><strong><a name="session">Session</a></strong>: handling sessions via cookies.
<hr/>

__<a name="type-sessionoptions">`SessionOptions`</a>__

|   Name    |                                                  Type                                                   |                 Description                  | Default |
| --------- | ------------------------------------------------------------------------------------------------------- | -------------------------------------------- | ------- |
| __keys*__ | <em>string[]</em>                                                                                       | A set of keys to be installed in `app.keys`. | -       |
| use       | <em>boolean</em>                                                                                        | Use this middleware for every request.       | `false` |
| config    | <em><a href="#type-sessionconfig" title="Configuration passed to `koa-session`.">SessionConfig</a></em> | `koa-session` configuration.                 | -       |
</summary>

__<a name="type-sessionconfig">`SessionConfig`</a>__: Configuration passed to `koa-session`.

|   Name    |              Type              |                                                                                        Description                                                                                        |  Default   |
| --------- | ------------------------------ | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ---------- |
| key       | <em>string</em>                | Cookie key.                                                                                                                                                                               | `koa:sess` |
| maxAge    | <em>(number \| 'session')</em> | maxAge in ms with default of 1 day. `session` will result in a cookie that expires when session/browser is closed. Warning: If a session cookie is stolen, this cookie will never expire. | `86400000` |
| overwrite | <em>boolean</em>               | Can overwrite or not.                                                                                                                                                                     | `true`     |
| httpOnly  | <em>boolean</em>               | httpOnly or not.                                                                                                                                                                          | `true`     |
| signed    | <em>boolean</em>               | Signed or not.                                                                                                                                                                            | `true`     |
| rolling   | <em>boolean</em>               | Force a session identifier cookie to be set on every response. The expiration is reset to the original maxAge, resetting the expiration countdown.                                        | `false`    |
| renew     | <em>boolean</em>               | Renew session when session is nearly expired, so we can always keep user logged in.                                                                                                       | `false`    |


</details>

<p align="center"><a href="#table-of-contents"><img src=".documentary/section-breaks/5.svg?sanitize=true" width="15"></a></p>
<a href="https://github.com/koa-modules/multer"><img src="https://raw.github.com/idiocc/core/master/images/multer.svg?sanitize=true" align="left" height="100"></a>
<details>
<summary><strong><a name="file-uploads">File Uploads</a></strong>: receiving files on the server.
<hr/>

__<a name="type-multeroptions">`MulterOptions`</a>__

|  Name  |                    Type                     |              Description               | Default |
| ------ | ------------------------------------------- | -------------------------------------- | ------- |
| use    | <em>boolean</em>                            | Use this middleware for every request. | `false` |
| config | <em>[MulterConfig](#type-multerconfig)</em> | `koa-multer` configuration.            | -       |
</summary>

[`import('http').IncomingMessage`](https://nodejs.org/api/http.html#http_class_http_incomingmessage) __<a name="type-httpincomingmessage">`http.IncomingMessage`</a>__

[`import('fs').Stats`](https://nodejs.org/api/fs.html#fs_class_fs_stats) __<a name="type-fsstats">`fs.Stats`</a>__

[`import('koa-multer').StorageEngine`](https://github.com/expressjs/multer#storage) __<a name="type-koa-multerstorageengine">`koa-multer.StorageEngine`</a>__

[`import('koa-multer').File`](https://github.com/expressjs/multer#file-information.) __<a name="type-koa-multerfile">`koa-multer.File`</a>__

__<a name="type-limits">`Limits`</a>__: [An object](https://github.com/expressjs/multer#limits) specifying the size limits.

|     Name      |      Type       |                                 Description                                  |  Default   |
| ------------- | --------------- | ---------------------------------------------------------------------------- | ---------- |
| fieldNameSize | <em>number</em> | Max field name size in bytes.                                                | `100`      |
| fieldSize     | <em>number</em> | Max field value size in bytes.                                               | `1024`     |
| fields        | <em>number</em> | Max number of non-file fields.                                               | `Infinity` |
| fileSize      | <em>number</em> | For multipart forms, the max file size in bytes.                             | `Infinity` |
| files         | <em>number</em> | For multipart forms, the max number of file fields.                          | `Infinity` |
| parts         | <em>number</em> | For multipart forms, the max number of parts (fields + files).               | `Infinity` |
| headerPairs   | <em>number</em> | For multipart forms, the max number of header key=&gt; value pairs to parse. | `2000`     |

__<a name="type-multerconfig">`MulterConfig`</a>__

|     Name     |                                                                  Type                                                                  |                                           Description                                           | Default |
| ------------ | -------------------------------------------------------------------------------------------------------------------------------------- | ----------------------------------------------------------------------------------------------- | ------- |
| dest         | <em>string</em>                                                                                                                        | Where to store the files.                                                                       | -       |
| storage      | <em>StorageEngine</em>                                                                                                                 | Where to store the files.                                                                       | -       |
| fileFilter   | <em>(req: IncomingMessage, file: File, callback: (error: (Error|null), acceptFile: boolean)) => void</em>                              | [Function](https://github.com/expressjs/multer#filefilter) to control which files are accepted. | -       |
| limits       | <em><a href="#type-limits" title="[An object](https://github.com/expressjs/multer#limits) specifying the size limits.">Limits</a></em> | Limits of the uploaded data.                                                                    | -       |

CHANGELOG.md
## 11 May 2019

### [1.4.0](https://github.com/idiojs/core/compare/v1.3.6...v1.4.0)

- [deps] Switch to [`@goa/koa`](https://github.com/idiocc/goa): The Koa web-server compiled using [Google Closure Compiler](https://github.com/google/closure-compiler/) via [Depack](https://artdecocode.com/depack/).

## 4 April 2019

### [1.3.6](https://github.com/idiojs/core/compare/v1.3.5...v1.3.6)

- [deps] Unfix dependencies.

## 2 March 2019

### 1.3.5

- [deps] Update dependencies (jsx+frontend: pass through `true` boolean attribute).

## 26 February 2019

### 1.3.4

- [deps] Update dependencies.

## 19 February 2019

### 1.3.3

- [deps] Update dependencies for _Frontend_ fix.

## 10 February 2019

### 1.3.2

- [feature] Update `frontend` to resolve entries and local deps in `node_modules`, serve linked packages, and simplify the source transforms.

## 8 February 2019

### 1.3.1

- [feature] Update `frontend` to serve direct imports (`import 'package'`).

## 7 February 2019

### 1.3.0

- [feature] Add the `frontend` middleware.

## 26 January 2019

### 1.2.0

- [feature] Return the `server`.

## 11 January 2018

### 1.1.0

- [feature] Cors middleware.
- [doc] Fix font copyright.
- [deps] Update dependencies and tests for newer `rqt.

## 25 September 2018

### 1.0.2

- [dep] Add types dependencies.

### 1.0.1

- [fix] Explicit return type.

### 1.0.0

- Release the first version.

## 4 July 2018

### 0.0.0

- Create `@idio/core` with [`mnp`][https://mnpjs.org]
- [repository]: `src`, `test`

src/index.js
import startApp from './lib/start-app'

/**
 * Start the server.
 * @param {MiddlewareConfig} [middlewareConfig] Middleware configuration for the `idio` `core` server.
 * @param {SessionOptions} [middlewareConfig.session] `session` options.
 * @param {MulterOptions} [middlewareConfig.multer] `multer` options.
 * @param {CSRFOptions} [middlewareConfig.csrf] `csrf` options.
 * @param {BodyparserOptions} [middlewareConfig.bodyparser] `bodyparser` options.
 * @param {CompressOptions} [middlewareConfig.compress] `compress` options.
 * @param {CheckauthOptions} [middlewareConfig.checkauth] `checkauth` options.
 * @param {LoggerOptions} [middlewareConfig.logger] `logger` options.
 * @param {StaticOptions} [middlewareConfig.static] `static` options.
 * @param {CorsOptions} [middlewareConfig.cors] `cors` options.
 * @param {FrontendOptions} [middlewareConfig.frontend] `frontend` options. If the option is specified, the middleware always will be used, i.e., no need to pass `use: true`.
 * @param {Config} [config] Server configuration object.
 * @param {number} [config.port=5000] The port on which to start the server. Default `5000`.
 * @param {string} [config.host="0.0.0.0"] The host on which to listen. Default `0.0.0.0`.
 * @example
```js
// start a server, and serve files from the "static" directory.
await idioCore({
  static: {
    use: true,
    root: 'static',
    config: {
      hidden: true,
    },
  },
})
```
 */
async function idioCore(middlewareConfig = {}, config = {}) {
  const res = await startApp(middlewareConfig, config)
  const { url, app, server, router, middleware } = res

  return { url, app, router, server, middleware }
}

// here until https://github.com/Microsoft/TypeScript/issues/26921 solved
/**
 * @typedef {import('@goa/koa').Middleware} _goa.Middleware
 */

/* documentary types/options/bodyparser.xml */
/**
 * @typedef {Object} BodyparserOptions
 * @prop {boolean} [use=false] Use this middleware for every request. Default `false`.
 * @prop {BodyparserConfig} [config] `koa-bodyparser` configuration.
 */

/* documentary types/config/bodyparser.xml */
/**
 * @typedef {import('@goa/koa').Context} Context
 *
 * @typedef {Object} BodyparserConfig
 * @prop {string[]} [enableTypes="['json', 'form']"] Parser will only parse when request type hits enableTypes. Default `['json', 'form']`.
 * @prop {string} [encode="utf-8"] Requested encoding. Default `utf-8`.
 * @prop {string} [formLimit="56kb"] Limit of the urlencoded body. If the body ends up being larger than this limit a 413 error code is returned. Default `56kb`.
 * @prop {string} [jsonLimit="1mb"] Limit of the json body. Default `1mb`.
 * @prop {boolean} [strict=true] When set to true, JSON parser will only accept arrays and objects. Default `true`.
 * @prop {(ctx: Context) => boolean} [detectJSON="null"] Custom json request detect function. Default `null`.
 * @prop {{json: string[], form: string[], text: string[]}} [extendTypes] Support extend types.
 * @prop {(err: Error, ctx: Context) => void} [onerror] Support custom error handle.
 */

/* documentary types/options/checkauth.xml */
/**
 * @typedef {Object} CheckauthOptions
 * @prop {boolean} [use=false] Use this middleware for every request. Default `false`.
 */

/**
 * @typedef {Object} CheckauthOptions
 * @prop {boolean} [use=false] Use this middleware for every request. Default `false`.
 */

/* documentary types/options/compress.xml */
/**
 * @typedef {Object} CompressOptions
 * @prop {boolean} [use=false] Use this middleware for every request. Default `false`.
 * @prop {CompressConfig} [config] `koa-compress` configuration.
 */

/* documentary types/config/compress.xml */
/**
 * @typedef {Object} CompressConfig
 * @prop {(content_type: string) => boolean} [filter] An optional function that checks the response content type to decide whether to compress. By default, it uses `compressible`.
 * @prop {number} [threshold=1024] Minimum response size in bytes to compress. Default `1024`.
 * @prop {number} [flush] Default: `zlib.constants.Z_NO_FLUSH`.
 * @prop {number} [finishFlush] Default: `zlib.constants.Z_FINISH`.
 * @prop {number} [chunkSize] Default: `16*1024`.
 * @prop {number} [windowBits] Support extend types.
 * @prop {number} [level] Compression only.
 * @prop {number} [memLevel] Compression only.
 * @prop {number} [strategy] Compression only.
 * @prop {*} [dictionary] Deflate/inflate only, empty dictionary by default.
 */

/* documentary types/options/csrf.xml */
/**
 * @typedef {Object} CSRFOptions
 * @prop {boolean} [use=false] Use this middleware for every request. Default `false`.
 * @prop {CSRFConfig} [config] `koa-csrf` configuration.
 */

/* documentary types/config/csrf.xml */
/**
 * @typedef {Object} CSRFConfig
 * @prop {string} [invalidSessionSecretMessage]
 * @prop {number} [invalidSessionSecretStatusCode]
 * @prop {string} [invalidTokenMessage]
 * @prop {number} [invalidTokenStatusCode]
 * @prop {string[]} [excludedMethods]
 * @prop {boolean} [disableQuery]
 */

/* documentary types/options/logger.xml */
/**
 * @typedef {Object} LoggerOptions
 * @prop {boolean} [use=false] Use this middleware for every request. Default `false`.
 * @prop {LoggerConfig} [config] `koa-logger` configuration.
 */

/* documentary types/config/logger.xml */
/**
 * @typedef {Object} LoggerConfig
 * @prop {(str: string, args: [string, string, string, string, string, string, string]) => void} [transporter] Param `str` is output string with ANSI Color, and you can get pure text with other modules like `strip-ansi`. Param `args` is a array by `[format, method, url, status, time, length]`.
 */

/* documentary types/options/multer.xml */
/**
 * @typedef {Object} MulterOptions
 * @prop {boolean} [use=false] Use this middleware for every request. Default `false`.
 * @prop {MulterConfig} [config] `koa-multer` configuration.
 */

/* documentary types/config/multer.xml */
/**
 * @typedef {import('http').IncomingMessage} IncomingMessage
 * @typedef {import('fs').Stats} Stats
 * @typedef {import('koa-multer').StorageEngine} StorageEngine
 * @typedef {import('koa-multer').File} File
 *
 * @typedef {Object} Limits [An object](https://github.com/expressjs/multer#limits) specifying the size limits.
 * @prop {number} [fieldNameSize=100] Max field name size in bytes. Default `100`.
 * @prop {number} [fieldSize=1024] Max field value size in bytes. Default `1024`.
 * @prop {number} [fields=Infinity] Max number of non-file fields. Default `Infinity`.
 * @prop {number} [fileSize=Infinity] For multipart forms, the max file size in bytes. Default `Infinity`.
 * @prop {number} [files=Infinity] For multipart forms, the max number of file fields. Default `Infinity`.
 * @prop {number} [parts=Infinity] For multipart forms, the max number of parts (fields + files). Default `Infinity`.
 * @prop {number} [headerPairs=2000] For multipart forms, the max number of header key=> value pairs to parse. Default `2000`.
 *
 * @typedef {Object} MulterConfig
 * @prop {string} [dest] Where to store the files.
 * @prop {StorageEngine} [storage] Where to store the files.
 * @prop {(req: IncomingMessage, file: File, callback: (error: Error | null, acceptFile: boolean)) => void} [fileFilter] [Function](https://github.com/expressjs/multer#filefilter) to control which files are accepted.
 * @prop {Limits} [limits] Limits of the uploaded data.
 * @prop {boolean} [preservePath=false] Keep the full path of files instead of just the base name. Default `false`.
 */

/* documentary types/options/session.xml */
/**
 * @typedef {Object} SessionOptions
 * @prop {string[]} keys A set of keys to be installed in `app.keys`.
 * @prop {boolean} [use=false] Use this middleware for every request. Default `false`.
 * @prop {SessionConfig} [config] `koa-session` configuration.
 */

/* documentary types/config/session.xml */
/**
 * @typedef {Object} SessionConfig Configuration passed to `koa-session`.
 * @prop {string} [key="koa:sess"] Cookie key. Default `koa:sess`.
 * @prop {number|'session'} [maxAge=86400000] maxAge in ms with default of 1 day. `session` will result in a cookie that expires when session/browser is closed. Warning: If a session cookie is stolen, this cookie will never expire. Default `86400000`.
 * @prop {boolean} [overwrite=true] Can overwrite or not. Default `true`.
 * @prop {boolean} [httpOnly=true] httpOnly or not. Default `true`.
 * @prop {boolean} [signed=true] Signed or not. Default `true`.
 * @prop {boolean} [rolling=false] Force a session identifier cookie to be set on every response. The expiration is reset to the original maxAge, resetting the expiration countdown. Default `false`.
 * @prop {boolean} [renew=false] Renew session when session is nearly expired, so we can always keep user logged in. Default `false`.
 */

/* documentary types/options/static.xml */
/**
 * @typedef {Object} StaticOptions
 * @prop {string|string[]} root Root or multiple roots from which to serve files.
 * @prop {boolean} [use=false] Use this middleware for every request. Default `false`.
 * @prop {string} [mount="/"] Path from which to serve files. Default `/`.
 * @prop {number} [maxage=0] How long to cache file for. Default `0`.
 * @prop {StaticConfig} [config] `koa-static` configuration.
 */

/* documentary types/config/static.xml */
/**
 * @typedef {import('http').ServerResponse} ServerResponse
 *
 * @typedef {(res: ServerResponse, path: string, stats: Stats) => any} SetHeaders
 *
 * @typedef {Object} StaticConfig
 * @prop {number} [maxage=0] Browser cache max-age in milliseconds. Default `0`.
 * @prop {boolean} [hidden=false] Allow transfer of hidden files. Default `false`.
 * @prop {string} [index="index.html"] Default file name. Default `index.html`.
 * @prop {boolean} [defer=false] If `true`, serves after return next(), allowing any downstream middleware to respond first. Default `false`.
 * @prop {boolean} [gzip=true] Try to serve the gzipped version of a file automatically when gzip is supported by a client and if the requested file with `.gz` extension exists. Default `true`.
 * @prop {boolean} [br=true] Try to serve the brotli version of a file automatically when brotli is supported by a client and if the requested file with `.br` extension exists (note, that brotli is only accepted over https). Default `true`.
 * @prop {SetHeaders} [setHeaders] Function to set custom headers on response.
 * @prop {boolean} [extensions=false] Try to match extensions from passed array to search for file when no extension is sufficed in URL. First found is served. Default `false`.
 */

/* documentary types/options/cors.xml */
/**
 * @typedef {import('koa').Context} Context
 *
 * @typedef {Object} CorsOptions
 * @prop {string|Array<string>|((ctx: Context) => string)} [origin] The origin or an array of origins to accept as valid. In case of an array, the origin from the request headers will be searched in the array, and if found, it will be returned (since browsers only support a single `Access-Control-Allow-Origin` header). If a function is passed, it should return the string with the origin to set. If not passed, the request origin is returned, allowing any origin to access the resource.
 * @prop {boolean} [use=false] Use this middleware for every request. Default `false`.
 * @prop {CorsConfig} [config] `@koa/cors` configuration.
 */

/* documentary types/config/cors.xml */
/**
 * @typedef {Object} CorsConfig
 * @prop {string} [origin="request Origin header"] `Access-Control-Allow-Origin` header value. Default `request Origin header`.
 * @prop {string|Array<string>} [allowMethods="GET,HEAD,PUT,POST,DELETE,PATCH"] `Access-Control-Allow-Methods` header value. Default `GET,HEAD,PUT,POST,DELETE,PATCH`.
 * @prop {string|Array<string>} [exposeHeaders] `Access-Control-Expose-Headers` header value.
 * @prop {string|Array<string>} [allowHeaders] `Access-Control-Allow-Headers` header value.
 * @prop {string|number} [maxAge] `Access-Control-Max-Age` header value in seconds.
 * @prop {boolean} [credentials=false] `Access-Control-Allow-Credentials` header value. Default `false`.
 * @prop {boolean} [keepHeadersOnError=false] Add set headers to `err.header` if an error is thrown. Default `false`.
 */

/* documentary types/middleware.xml */
/**
 * @typedef {Object} MiddlewareConfig Middleware configuration for the `idio` `core` server.
 * @prop {SessionOptions} [session] `session` options.
 * @prop {MulterOptions} [multer] `multer` options.
 * @prop {CSRFOptions} [csrf] `csrf` options.
 * @prop {BodyparserOptions} [bodyparser] `bodyparser` options.
 * @prop {CompressOptions} [compress] `compress` options.
 * @prop {CheckauthOptions} [checkauth] `checkauth` options.
 * @prop {LoggerOptions} [logger] `logger` options.
 * @prop {StaticOptions} [static] `static` options.
 * @prop {CorsOptions} [cors] `cors` options.
 * @prop {FrontendOptions} [frontend] `frontend` options. If the option is specified, the middleware always will be used, i.e., no need to pass `use: true`.
 */

/* documentary types/config/frontend.xml */
/**
 * @typedef {Object} FrontendConfig
 * @prop {string} [pragma="import { h } from 'preact'"] The pragma function to import. This enables to skip writing `h` at the beginning of each file. JSX will be transpiled to have `h` pragma, therefore to use React it's possible to do `import { createElement: h } from 'react'`. Default `import { h } from 'preact'`.
 */

/* documentary types/options/frontend.xml */
/**
 * @typedef {Object} FrontendOptions Allows to serve front-end JS files and CSS as modules, including from node_modules folder.
 * @prop {(string|Array<string>)} [directory="frontend"] The directory or directories from which to serve files. Default `frontend`.
 * @prop {FrontendConfig} [config] `@idio/frontend` configuration.
 */

/* documentary types/config.xml */
/**
 * @typedef {Object} Config Server configuration object.
 * @prop {number} [port=5000] The port on which to start the server. Default `5000`.
 * @prop {string} [host="0.0.0.0"] The host on which to listen. Default `0.0.0.0`.
 */

export default idioCore